flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/HadoopDataInputStream.java

```diff
@@ -140,4 +144,43 @@ public void skipFully(long bytes) throws IOException {
             bytes -= fsDataInputStream.skip(bytes);
         }
     }
+
+    @Override
+    public int read(ByteBuffer byteBuffer) throws IOException {
+        // Not all internal stream supports ByteBufferReadable
+        if (fsDataInputStream.hasCapability(StreamCapabilities.READBYTEBUFFER)) {
+            return fsDataInputStream.read(byteBuffer);
+        } else {
+            // Fallback to read byte then put
```

Comment:
Why is this fallback part different from the one in `public int read(long position, ByteBuffer byteBuffer)`

Reply:
Since position read may fallback to just use the method of position read with extra byte array in hadoop, I'd prefer to implement this logic without extra array overhead.